### PR TITLE
unixODBCDrivers.msodbcsql18: init at 18.1.1.1

### PR DESCRIPTION
**Pull request by SamirTalwar**
###### Description of changes

This is the ODBC Driver 18 for SQL Server, which now supports arm64 (and therefore modern Mac hardware). Download links were acquired from the [Linux](https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server?view=sql-server-ver16) and [macOS](https://learn.microsoft.com/en-us/sql/connect/odbc/linux-mac/install-microsoft-odbc-driver-sql-server-macos?view=sql-server-ver16) installation pages.

I used the Debian packages, as that's what the `msodbcsql17` package uses.

I have verified these build appropriately. I haven't tested them thoroughly in nixpkgs, but I have tested them at length as part of [hasura/graphql-engine](https://github.com/hasura/graphql-engine). The contents are copied from [_nix/msodbcsql18-linux.nix_](https://github.com/hasura/graphql-engine/blob/c3afa0fdd7eece8eb1bc708e4e241efe36c18618/nix/msodbcsql18-linux.nix) and [_nix/msodbcsql18-darwin.nix_](https://github.com/hasura/graphql-engine/blob/c3afa0fdd7eece8eb1bc708e4e241efe36c18618/nix/msodbcsql18-darwin.nix).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by willbush** [comment]
Anything I can do to help get this done?

I have a rust project at work that uses odbc-api to connect to mssql. I recently updated the `flake.nix` for that project which follows unstable nixpkgs, and I'm getting this message:

>OpenSSL 1.1 is reaching its end of life on 2023/09/11 and cannot be supported through the NixOS 23.05 release cycle. https://www.openssl.org/blog/blog/2023/03/28/1.1.1-EOL/

Note that ODBC Driver 17 for SQL Server depends on OpenSSL 1.0 or 1.1:

>[Microsoft][ODBC Driver 17 for SQL Server]SSL Provider: [OpenSSL library could not be loaded, make sure OpenSSL 1.0 or 1.1 is installed]

The temp work around is simple enough:

```
nixpkgs.config.permittedInsecurePackages = [
  "openssl-1.1.1t"
];
```

I'm guessing msodbcsql18 is the way forward.

**Comment by SamirTalwar** [comment]
I'm not sure, but I think reviewing the PR according to the [reviewing contributions](https://nixos.org/manual/nixpkgs/stable/#chap-reviewing-contributions) documentation would help.

I'm happy to get this PR up to date with the latest minor version (assuming one's been published since I opened it) and fix the merge conflict if it looks like there's good odds on merging it.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:184]: Could you try to replace this `if-else` with an expression common for both platforms? Many attributes, (especially `meta`) is common to all platforms...

**Comment by doronbehar** [comment]
Also there is a merge conflict now.

**Comment by SamirTalwar** [comment]
@doronbehar, I can think of a few ways to remove the duplication. What would you prefer?

1. Add a `if isDarwin then … else …` block for each attribute. There are 5 attributes that would need this. (FWIW I started doing this and it seems to result in _more_ duplication.)
2. Pull out two functions that each takes the common data and then adds OS-specific attributes, then call the correct one.
3. Keep the variants separate but move them to top-level attributes (e.g. `msodbcsql18Linux`) and then just alias the correct one to `msodbcsql18`.

There might be a better way that I am not aware of.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix:185]: @doronbehar, I can think of a few ways to remove the duplication. What would you prefer?

1. Add a `if isDarwin then … else …` block for each attribute. There are 5 attributes that would need this. (FWIW I started doing this and it seems to result in _more_ duplication.)
2. Pull out two functions that each takes the common data and then adds OS-specific attributes, then call the correct one.
3. Keep the variants separate but move them to top-level attributes (e.g. `msodbcsql18Linux`) and then just alias the correct one to `msodbcsql18`.

There might be a better way that I am not aware of.

Comment (doronbehar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix:185]: > 1. ... FWIW I started doing this and it seems to result in _more_ duplication.

Sounds surprising... Sparing the `meta` duplication seems to win this IMO.

> 2. ...

That's too complicated in my opinion.

> 3. ...

No no no.

**Comment by doronbehar** [comment]
Oh and next time, when you fix a merge conflict, `git rebase master`, and don't merge master to your branch.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Alright, done. I also rebased and squashed.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:269]: This can be squashed into:

```suggestion
        driver = "lib/libmsodbcsql.${versionMajor}.${if stdenv.isDarwin the "dylib" else "${versionMinor}.so.${versionAdditional}"}"
```

That `versionAdditional` in the end needs to be validated on a Linux machine...

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:276]:
```suggestion
        platforms = platforms.unix
```

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [inline pkgs/development/libraries/unixODBCDrivers/default.nix:263]: What is this needed for?

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix:263]: It's useful when building upon it to generate the `odbcinst.ini` file. [Here's an example.](https://github.com/hasura/graphql-engine/blob/47baf4addc33f5f88ecc2309691418a6753a2b11/nix/shell.nix#L20-L32)

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix:184]: I can probably arrange this but note that `nix-update` is unlikely to work here. The URLs for macOS include a content hash and will therefore change completely.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: I prefer this too; cheers.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: I would prefer to keep it as-is; I find staggered conditions like this, where something inside the block is indented less than the thing outside the block, quite hard to read.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Good tip, cheers.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: I like this.

Happy to use `hash = "sha256:..."`; I wasn't aware of it. I kept the base-64 encoding because that's what `nix-prefetch-url` gives me.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Done!

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: > I've done this but note that `nix-update` is unlikely to work here. The URLs for macOS include a content hash and will therefore change completely.
> 
> `msodbcsql17` should probably change too, but as I don't use it, I don't really want to touch it.

Not sure what do you mean by "_a content hash_", but I agree that updating all of the hashes at once is currently impossible for `nix-update`, hence this shouldn't be a consideration.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: > It's useful when building upon it to generate the `odbcinst.ini` file. [Here's an example.](https://github.com/hasura/graphql-engine/blob/47baf4addc33f5f88ecc2309691418a6753a2b11/nix/shell.nix#L20-L32)

It'd be nice to comment something near this, near the `passthru` value...

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: > I would prefer to keep it as-is; I find staggered conditions like this, where something inside the block is indented less than the thing outside the block, quite hard to read.

Within reasonably common & sensible coding style and coherence, it is your right to use whatever indentation and coding style you choose for packages, if you are consistent. The advantage of the indentation I suggested, is that it minimizes diffs in case lines from one platform are moved to other platforms / conditions etc.

**Comment by doronbehar** [comment]
Result of `nixpkgs-review pr 197607` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>unixODBCDrivers.msodbcsql18</li>
  </ul>
</details>

**Comment by doronbehar** [comment]
I have no idea what is the purpose of this package, but it seems to evaluate and build fine. I will merge after the comment near the passthru is finished, and optionally the indentation / coding style is fixed.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (doronbehar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Quoting myself:

> That `versionAdditional` in the end needs to be validated on a Linux machine...

Seems fine.

Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: Yeah, I tested it myself too.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: It's the same explanation for all ODBC drivers, so I added a long comment to the top of the file and short comments near each `passthru`.

File: pkgs/development/libraries/unixODBCDrivers/default.nix
Comment (SamirTalwar) [reply pkgs/development/libraries/unixODBCDrivers/default.nix]: A tough one: one's easier when reading code, one's easier when reading a diff.

I will keep it as it is but I totally understand why you have a different preference.

**Comment by doronbehar** [comment]
This almost ready @SamirTalwar, one last thing: Rewrite please the Git History as follows:

```
maintainers: add SamirTalwar
unixODBCDrivers.msodbcsql18: init at 18.1.1.1
unixODBCDrivers: Document `fancyName` and `driver`
```

**Comment by SamirTalwar** [comment]
History rewritten! Thanks, @doronbehar.

**Review by doronbehar — APPROVED** [review]

**Comment by SamirTalwar** [comment]
Thanks for merging this!

I will probably submit a PR to update the version at some point, but I'll have to test it out at work first.

**Comment by jrseliga** [comment]
@SamirTalwar hope I'm not too unfairly hijacking this PR.

I've installed `unixODBCDrivers.msodbcsql18` and `php80Extensions.pdo_sqlsrv`. However, when attempting to make a connection to the SQL Server I receive:

```
This extension requires the Microsoft ODBC Driver for SQL Server to communicate with SQL Server.
Access the following URL to download the ODBC Driver for SQL Server for x64: https://go.microsoft.com/fwlink/?LinkId=163712
```

Is it possible `msodbcsql18` is not being properly registered in `obdcinst.ini` after install?

I'm extremely new to the Nix world, but it looks like there is a [mechanism](https://github.com/NixOS/nixpkgs/blob/75565166d67449ecfbb132d81da9e3a4bef29dc3/nixos/modules/config/unix-odbc-drivers.nix#L24C1-L27) intended for `unixODBCDrivers` to be registered in `odbcinst.ini`?

**Comment by willbush** [comment]
@jrseliga I used to do this in a `flake.nix` to get around the issue. The versions below will have to be updated.

```nix
shellHook = ''
  # see https://www.systutorials.com/docs/linux/man/7-unixODBC/
  # Overloads path to unixODBC configuration files. By default equals to '/etc'.
  export ODBCSYSINI=$(realpath ./)

  echo "[ODBC Driver 17 for SQL Server]" > ./odbcinst.ini
  echo "Description = ODBC Driver 17 for SQL Server" >> ./odbcinst.ini
  echo "Driver = ${pkgs.unixODBCDrivers.msodbcsql17}/lib/libmsodbcsql-17.7.so.1.1" >> ./odbcinst.ini
'';
```

**Comment by SamirTalwar** [comment]
FWIW, the `OSBCSYSINI` environment variable had mixed results when we tried it. It worked on some machines but not others.

[We ended up building the path into the executable.](https://github.com/hasura/graphql-engine/blob/76604c6f231fec55d46c55f5dc60f3bca4987547/nix/shell.nix#L20-L36)

**Pull request merged**
